### PR TITLE
Adding Hbase-2.x modules.

### PR DESCRIPTION
**Pull request by spollapally**
Sorry for this large commit. A lot of it is moving existing code/config around, and test cases in bigtable-hbase-integration-tests-2.x mostly remain same from 1.x version.

### CODE REVIEW

**Comment by sduskis** [comment]
Can you please add a list of files and specifics to look at in this PR?  What did you spend time on that we should pay close attention to as reviewers?

**Comment by spollapally** [comment]
Review notes - Files listed under "Changed" header below have updates, in many cases minor, except for new classes like 2.x versions of BigtableAdmin.java and BigtableConnection.java.

I created tag "2x_base_copy" in my clone (https://github.com/spollapally/cloud-bigtable-client/tree/2x_base_copy) which makes the comparison far more easier to comprehend (at least in eclipse. github "compare" didn't do that well).  This tag is the initial copy over from corresponding 1x modules.

## Changed
pom.xml
bigtable-hbase-parent/pom.xml

bigtable-hbase-parent/bigtable-hbase-2.x/pom.xml
bigtable-hbase-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAdmin.java
bigtable-hbase-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableConnection.java

bigtable-hbase-parent/bigtable-hbase-2.x-shaded/pom.xml
bigtable-hbase-parent/bigtable-hbase-2.x/pom.xml
bigtable-hbase-parent/bigtable-hbase-integration-tests-2.x/pom.xml

bigtable-hbase-parent/bigtable-hbase-integration-tests-2.x/src/test/java/com/google/cloud/bigtable/hbase/TestAppend.java
bigtable-hbase-parent/bigtable-hbase-integration-tests-2.x/src/test/java/com/google/cloud/bigtable/hbase/TestBasicOps.java
bigtable-hbase-parent/bigtable-hbase-integration-tests-2.x/src/test/java/com/google/cloud/bigtable/hbase/TestCheckAndMutate.java
bigtable-hbase-parent/bigtable-hbase-integration-tests-2.x/src/test/java/com/google/cloud/bigtable/hbase/TestCreateTable.java
bigtable-hbase-parent/bigtable-hbase-integration-tests-2.x/src/test/java/com/google/cloud/bigtable/hbase/TestPut.java

## Pretty much intact

bigtable-hbase-parent/bigtable-hbase-2.x-hadoop/src/main/java/com/google/cloud/bigtable/hbase/MavenPlaceholder.java
bigtable-hbase-parent/bigtable-hbase-2.x-hadoop/src/test/java/com/google/cloud/bigtable/hbase/TestBigtableConnection.java
bigtable-hbase-parent/bigtable-hbase-2.x-hadoop/src/test/resources/log4j.properties

bigtable-hbase-parent/bigtable-hbase-2.x-shaded/src/main/java/com/google/cloud/bigtable/hbase/MavenPlaceholder.java

bigtable-hbase-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAdmin.java
bigtable-hbase-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableConnection.java
bigtable-hbase-parent/bigtable-hbase-2.x/src/main/resources/log4j.properties
bigtable-hbase-parent/bigtable-hbase-2.x/src/test/java/com/google/cloud/bigtable/hbase/ManyThreadDriver.java
bigtable-hbase-parent/bigtable-hbase-2.x/src/test/java/com/google/cloud/bigtable/hbase/RecreateConnectionDriver.java
bigtable-hbase-parent/bigtable-hbase-2.x/src/test/java/com/google/cloud/bigtable/hbase/TestBigtableConnection.java

bigtable-hbase-parent/bigtable-hbase-integration-tests-2.x/src/main/java/com/google/cloud/bigtable/hbase/MavenPlaceholder.java
bigtable-hbase-parent/bigtable-hbase-integration-tests-2.x/src/test/java/com/google/cloud/bigtable/hbase/AbstractTest.java
bigtable-hbase-parent/bigtable-hbase-integration-tests-2.x/src/test/java/com/google/cloud/bigtable/hbase/DataGenerationHelper.java
bigtable-hbase-parent/bigtable-hbase-integration-tests-2.x/src/test/java/com/google/cloud/bigtable/hbase/EmulatorTests.java
bigtable-hbase-parent/bigtable-hbase-integration-tests-2.x/src/test/java/com/google/cloud/bigtable/hbase/IntegrationTests.java
bigtable-hbase-parent/bigtable-hbase-integration-tests-2.x/src/test/java/com/google/cloud/bigtable/hbase/IntegrationTestsGap.java
bigtable-hbase-parent/bigtable-hbase-integration-tests-2.x/src/test/java/com/google/cloud/bigtable/hbase/IntegrationTestsNoKnownGap.java
bigtable-hbase-parent/bigtable-hbase-integration-tests-2.x/src/test/java/com/google/cloud/bigtable/hbase/KnownEmulatorGap.java
bigtable-hbase-parent/bigtable-hbase-integration-tests-2.x/src/test/java/com/google/cloud/bigtable/hbase/KnownGap.java
bigtable-hbase-parent/bigtable-hbase-integration-tests-2.x/src/test/java/com/google/cloud/bigtable/hbase/Logger.java
bigtable-hbase-parent/bigtable-hbase-integration-tests-2.x/src/test/java/com/google/cloud/bigtable/hbase/SimpleIntegrationTests.java
bigtable-hbase-parent/bigtable-hbase-integration-tests-2.x/src/test/java/com/google/cloud/bigtable/hbase/TestBatch.java
bigtable-hbase-parent/bigtable-hbase-integration-tests-2.x/src/test/java/com/google/cloud/bigtable/hbase/TestBufferedMutator.java
bigtable-hbase-parent/bigtable-hbase-integration-tests-2.x/src/test/java/com/google/cloud/bigtable/hbase/TestColumnFamilyAdmin.java
bigtable-hbase-parent/bigtable-hbase-integration-tests-2.x/src/test/java/com/google/cloud/bigtable/hbase/TestDelete.java
bigtable-hbase-parent/bigtable-hbase-integration-tests-2.x/src/test/java/com/google/cloud/bigtable/hbase/TestDisableTable.java
bigtable-hbase-parent/bigtable-hbase-integration-tests-2.x/src/test/java/com/google/cloud/bigtable/hbase/TestDurability.java
bigtable-hbase-parent/bigtable-hbase-integration-tests-2.x/src/test/java/com/google/cloud/bigtable/hbase/TestFilters.java
bigtable-hbase-parent/bigtable-hbase-integration-tests-2.x/src/test/java/com/google/cloud/bigtable/hbase/TestGet.java
bigtable-hbase-parent/bigtable-hbase-integration-tests-2.x/src/test/java/com/google/cloud/bigtable/hbase/TestGetTable.java
bigtable-hbase-parent/bigtable-hbase-integration-tests-2.x/src/test/java/com/google/cloud/bigtable/hbase/TestImport.java
bigtable-hbase-parent/bigtable-hbase-integration-tests-2.x/src/test/java/com/google/cloud/bigtable/hbase/TestIncrement.java
bigtable-hbase-parent/bigtable-hbase-integration-tests-2.x/src/test/java/com/google/cloud/bigtable/hbase/TestListTables.java
bigtable-hbase-parent/bigtable-hbase-integration-tests-2.x/src/test/java/com/google/cloud/bigtable/hbase/TestPutGetDelete.java
bigtable-hbase-parent/bigtable-hbase-integration-tests-2.x/src/test/java/com/google/cloud/bigtable/hbase/TestScan.java
bigtable-hbase-parent/bigtable-hbase-integration-tests-2.x/src/test/java/com/google/cloud/bigtable/hbase/TestSingleColumnValueFilter.java
bigtable-hbase-parent/bigtable-hbase-integration-tests-2.x/src/test/java/com/google/cloud/bigtable/hbase/TestSnapshots.java
bigtable-hbase-parent/bigtable-hbase-integration-tests-2.x/src/test/java/com/google/cloud/bigtable/hbase/TestTimestamp.java
bigtable-hbase-parent/bigtable-hbase-integration-tests-2.x/src/test/java/com/google/cloud/bigtable/hbase/TestTruncateTable.java
bigtable-hbase-parent/bigtable-hbase-integration-tests-2.x/src/test/java/com/google/cloud/bigtable/hbase/test_env/BigtableEnv.java
bigtable-hbase-parent/bigtable-hbase-integration-tests-2.x/src/test/java/com/google/cloud/bigtable/hbase/test_env/MiniClusterEnv.java
bigtable-hbase-parent/bigtable-hbase-integration-tests-2.x/src/test/java/com/google/cloud/bigtable/hbase/test_env/SharedTestEnv.java
bigtable-hbase-parent/bigtable-hbase-integration-tests-2.x/src/test/java/com/google/cloud/bigtable/hbase/test_env/SharedTestEnvRule.java
bigtable-hbase-parent/bigtable-hbase-integration-tests-2.x/src/test/resources/hbase-site.xml
bigtable-hbase-parent/bigtable-hbase-integration-tests-2.x/src/test/resources/log4j.properties

File: bigtable-hbase-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAdmin.java [hunk 1]
```diff
@@ -0,0 +1,618 @@
...
+      throws IOException {
+    createTable((HTableDescriptor) desc, startKey, endKey, numRegions);
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public Future<Void> createTableAsync(TableDescriptor desc, byte[][] splitKeys)
+      throws IOException {
+    createTableAsync((HTableDescriptor) desc, splitKeys);
+    // Consider better options after adding support for async hbase2
```
Comment (sduskis) [inline bigtable-hbase-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAdmin.java]: Please add "TODO:" in front of any comments like this, where an action is required.

File: bigtable-hbase-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAdmin.java [hunk 2]
```diff
@@ -0,0 +1,618 @@
...
+    deleteNamespace(name);
+    // Consider better options after adding support for async hbase2
+    return CompletableFuture.runAsync(() -> {});
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public Future<Void> disableTableAsync(TableName tableName) throws IOException {
+    disableTable(tableName);
+    // Consider better options after adding support for async hbase2
```
Comment (sduskis) [inline bigtable-hbase-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAdmin.java]: TODO:

File: bigtable-hbase-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAdmin.java [hunk 3]
```diff
@@ -0,0 +1,618 @@
...
+    return CompletableFuture.runAsync(() -> {});
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public Future<Void> enableTableAsync(TableName tableName) throws IOException {
+    enableTable(tableName);
+    // Consider better options after adding support for async hbase2
```
Comment (sduskis) [inline bigtable-hbase-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAdmin.java]: TODO

File: bigtable-hbase-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAdmin.java [hunk 4]
```diff
@@ -0,0 +1,618 @@
...
+    throw new UnsupportedOperationException("modifyTable"); // TODO
+  }
+
+  @Override
+  public Future<Void> modifyTableAsync(TableDescriptor arg0) throws IOException {
+    throw new UnsupportedOperationException("modifyTableAsync"); // TODO
+  }
+
+  @Override
+  public Future<Void> modifyTableAsync(TableName arg0, TableDescriptor arg1) throws IOException {
```
Comment (sduskis) [inline bigtable-hbase-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAdmin.java]: Please add a TODO to implement all modifyTable varieties.

File: bigtable-hbase-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAdmin.java [hunk 5]
```diff
@@ -0,0 +1,618 @@
...
+    throw new UnsupportedOperationException("mergeRegionsAsync"); // TODO
+  }
+
+  @Override
+  public Future<Void> mergeRegionsAsync(byte[] arg0, byte[] arg1, boolean arg2) throws IOException {
+    throw new UnsupportedOperationException("mergeRegionsAsync"); // TODO
+  }
+
+  @Override
+  public Future<Void> modifyColumnFamilyAsync(TableName arg0, ColumnFamilyDescriptor arg1)
```
Comment (sduskis) [inline bigtable-hbase-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAdmin.java]: Please add a TODO here.  This should be implementable

File: bigtable-hbase-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAdmin.java [hunk 6]
```diff
@@ -0,0 +1,618 @@
...
+    throw new UnsupportedOperationException("splitOrMergeEnabledSwitch"); // TODO
+  }
+
+  @Override
+  public Future<Void> splitRegionAsync(byte[] arg0, byte[] arg1) throws IOException {
+    throw new UnsupportedOperationException("splitRegionAsync"); // TODO
+  }
+
+  @Override
+  public Future<Void> truncateTableAsync(TableName arg0, boolean arg1) throws IOException {
```
Comment (sduskis) [inline bigtable-hbase-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAdmin.java]: This is implementable.  Please add a TODO

File: bigtable-hbase-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableConnection.java [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+  public BigtableConnection(Configuration conf, ExecutorService pool, User user)
+      throws IOException {
+    super(conf);
+  }
+
+  protected SampledRowKeysAdapter createSampledRowKeysAdapter(TableName tableName,
+      ServerName serverName) {
+    return new SampledRowKeysAdapter(tableName, serverName) {
+      @Override
+      public HRegionLocation createHRegionLocation(HRegionInfo hRegionInfo, ServerName serverName) {
```
Comment (sduskis) [inline bigtable-hbase-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableConnection.java]: Please add a comment here about the fact that `HRegionLocation` is binary incompatible between hbase 1.x and 2.x

File: bigtable-hbase-parent/bigtable-hbase-integration-tests-2.x/src/test/java/com/google/cloud/bigtable/hbase/TestPut.java [hunk 8]
```diff
@@ -0,0 +1,398 @@
...
+    put.addColumn(COLUMN_FAMILY, goodQual, goodValue);
+    put.addColumn(badfamily, badQual, badValue);
+    NoSuchColumnFamilyException thrownException = null;
+    try {
+      table.put(put);
+    } catch (NoSuchColumnFamilyException e) {
+      thrownException = e;
+    }
+    Assert.assertNotNull("Exception should have been thrown", thrownException);
+    //Assert.assertEquals("Expecting one exception", 1, thrownException.getNumExceptions());
```
Comment (sduskis) [inline bigtable-hbase-parent/bigtable-hbase-integration-tests-2.x/src/test/java/com/google/cloud/bigtable/hbase/TestPut.java]: Please add a TODO here to investigate why this doesn't work.

**Comment by spollapally** [comment]
Updated the commit with TODOs and additional documentation.

**Comment by sduskis** [comment]
Looks good.  Please Squash and Merge.

**Comment by gamorris** [comment]
fyi - It looks like the new bigtable-hbase-integration-tests-2.x fail. We're seeing this in one CI environment:
```
Tests run: 115, Failures: 0, Errors: 1, Skipped: 1, Time elapsed: 71.604 sec <<< FAILURE! - in com.google.cloud.bigtable.hbase.IntegrationTestsNoKnownGap
testTableNames(com.google.cloud.bigtable.hbase.TestCreateTable)  Time elapsed: 0.015 sec  <<< ERROR!
java.io.IOException: Failed to create table 'aྜ쟖멉暢訋ⳑ詃ꐩ琐힆'
	at com.google.cloud.bigtable.hbase.TestCreateTable.testTableNames(TestCreateTable.java:107)
Caused by: com.google.bigtable.repackaged.io.grpc.StatusRuntimeException: INVALID_ARGUMENT: Invalid id for collection tables : Should match [_a-zA-Z0-9][-_.a-zA-Z0-9]* but found 'aྜ쟖멉暢訋ⳑ詃ꐩ琐힆'
	at com.google.cloud.bigtable.hbase.TestCreateTable.testTableNames(TestCreateTable.java:107)
```

Looks like the new test was copied from the old test... Was the old implementation doing some client-side validation? Or otherwise translating the errors? It appears to be getting the grpc `StatusRuntimeException` now instead of an `IllegalArgumentException`.

**Comment by sduskis** [comment]
That test seems to be flaky.  I'll turn off tests for hbase2 for now

**Comment by spollapally** [comment]
Committing a fix. The error is due to randomly generated table name which
server side doesn't like. A typo in my earlier commit was partly the issue.

**Pull request merged**